### ISSUE TITLE
`"contour"`

### ISSUE THREAD

**Issue by grantmcdermott** [grantmcdermott/tinyplot#218]

**Comment by grantmcdermott** [comment]
We could use `graphics::contourLines` to calculate the appropriate breaks, followed by some internal data reconstruction to grab the group IDs. 

Quick proof of concept that also uses that 'nested' ID coloring approach mentioned in https://github.com/grantmcdermott/tinyplot/issues/41#issuecomment-2322220131 (contour plots are another case where this nesting would be necessary).

``` r
library(tinyplot)

cl = contourLines(z = volcano)
cl = lapply(cl, as.data.frame)
cl = do.call(rbind, unname(Map(cbind, id = seq_along(cl), cl)))
head(cl)
#>   id level         x         y
#> 1  1   100 1.0000000 0.5515289
#> 2  1   100 0.9883721 0.5515289
#> 3  1   100 0.9778109 0.5666667
#> 4  1   100 0.9767442 0.5681956
#> 5  1   100 0.9651163 0.5681956
#> 6  1   100 0.9545551 0.5833333

cols = with(cl, tapply(factor(level), id, FUN = `[[`, 1))  # grab group colours
plt(
    y ~ x | id, cl,
    type = "l",
    col = hcl.colors(length(unique(cols)))[cols],
    legend = FALSE
)
#> Warning in tinyplot.default(x = x, y = y, by = by, facet = facet, facet.args = facet.args, : 
#> Continuous legends not supported for this plot type. Reverting to discrete legend.
```

![](https://i.imgur.com/rF3kkZu.png)

Filled contour plots would take a bit more thinking; the below doesn't work properly because of sub-optimal joins of the 'outer' polygons that should really extend into the plot margin. Also: each polygon should be drawn in 'donut' fashion&mdash;with the hole determined by the next contour&mdash;rather than filling in all the interior space.
``` r
plt(
    y ~ x | id, cl,
    type = "polypath", #rule = 'evenodd',
    col = hcl.colors(length(unique(cols)))[cols],
    legend = FALSE, fill = 0.1
)
#> Warning in tinyplot.default(x = x, y = y, by = by, facet = facet, facet.args = facet.args, : 
#> Continuous legends not supported for this plot type. Reverting to discrete legend.
```

![](https://i.imgur.com/nuEH57n.png)

<sup>Created on 2024-09-16 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Comment by grantmcdermott** [comment]
All valid points! I haven't had time to think about it much, just wanted to jot down a quick proof of concept. You're absolutely right that the implementation would require some thought w.r.t. user interface. (Doesn't lattice have a contour method that might give us some ideas for potential formula logic?)

**Comment by zeileis** [comment]
Good point. In `lattice` these are called `levelplot()` (for heatmaps) and `contourplot()` (for contour lines, with or without filling). Both of them are designed for 3D data and expect formulas like `z ~ x + y`. A third function with the same kind of interface is `wireframe()` which is the `lattice` version of `persp()`.

For the `volcano` data all three functions can be used like this on a matrix/table:

```
levelplot(volcano)
```

Or you can set up a data frame with three variables like this:

```
volcano2 <- data.frame(
  row = as.vector(row(volcano)),
  column = as.vector(col(volcano)),
  height = as.vector(volcano)
)
```

And then you can use the formula interface:

```
levelplot(height ~ row + column, data = volcano2)
```

If you additionally set `aspect = "iso"` you also get the same aspect ratio as in the matrix interface.

In short, I guess this is closer to what you had in mind, right? The plots I was looking for (smoothed scatterplots for 2d data) could then possibly be accomplished by deriving the density estimate and then calling the 3d functionality.

**Comment by zeileis** [comment]
Just a link so that I don't forget. It's not about contours but about heatmaps which we also mentioned above. Maybe this is useful for inspiration (maybe not): <https://jbengler.github.io/tidyheatmaps/>
